internal/storage/multi_store.go

```diff
@@ -0,0 +1,445 @@
+		id, err := store.Lookup(id)
+		if err != nil || id == "" {
+			continue
+		}
+		return store.Metadata(id)
+	}
+	return "", fmt.Errorf("Metadata for %s not found", id)
+}
+
+// DeleteContainer removes the specified container and its layer.
+func (s *multiStore) DeleteContainer(id string) error {
+	s.iterator.initialize()
+	for store := s.iterator.next(); store != nil; store = s.iterator.next() {
+		id, err := store.Lookup(id)
+		if err != nil || id == "" {
+			continue
+		}
+		return store.DeleteContainer(id)
+	}
+	return fmt.Errorf("container %s not found", id)
+}
+
+// DeleteImage removes the specified image if it is not referred to by any containers.
+func (s *multiStore) DeleteImage(id string, commit bool) (layers []string, err error) {
+	s.iterator.initialize()
+	for store := s.iterator.next(); store != nil; store = s.iterator.next() {
+		id, err := store.Lookup(id)
+		if err != nil || id == "" {
+			continue
+		}
+		return store.DeleteImage(id, commit)
+	}
+	return []string{}, fmt.Errorf("image %s not found", id)
+}
+
+// GetStore returns all the stores.
+func (s *multiStore) GetStore() map[string]cstorage.Store {
+	return s.store
+}
+
+func (s *multiStore) RunRoot() string {
+	return s.runRoot
+}
+
+func (s *multiStore) GraphRoot() string {
+	return s.graphRoot
+}
+
+// GetStoreForContainer returns the store for the given id or name.
+func (s *multiStore) GetStoreForContainer(idOrName string) (cstorage.Store, error) {
+	logrus.Debugf("GetStore for container %s", idOrName)
+	s.iterator.initialize()
+	for store := s.iterator.next(); store != nil; store = s.iterator.next() {
+		if _, err := store.Container(idOrName); err != nil {
+			continue
+		}
+		return store, nil
+	}
+	return nil, fmt.Errorf("error locating container %s", idOrName)
+}
+
+// multiStoreServer contains the store image servers based on the configured storage drivers.
+type multiStoreServer struct {
+	store      map[string]ImageServer
+	multiStore MultiStore
+	iterator   iteratorMultiStoreServer
+}
+
+// iteratorMultiStoreServer iterates over the multiStoreServer starting by the default storage driver.
+type iteratorMultiStoreServer struct {
+	keys             []string
+	multiStoreServer *multiStoreServer
+	counter          int
+}
+
+func createMultiStoreServerIterator(s *multiStoreServer) iteratorMultiStoreServer {
+	defaultStorage := s.multiStore.GetDefaultStorageDriver()
+	keys := []string{defaultStorage}
+	for k := range s.store {
+		if k == defaultStorage {
+			continue
+		}
+		keys = append(keys, k)
+	}
+	return iteratorMultiStoreServer{
+		multiStoreServer: s,
+		keys:             keys,
+	}
+}
+
+func (i *iteratorMultiStoreServer) initialize() {
+	i.counter = 0
+}
+
+func (i *iteratorMultiStoreServer) next() ImageServer {
+	if i.counter > len(i.multiStoreServer.store)-1 {
+		return nil
+	}
+	if e, ok := i.multiStoreServer.store[i.keys[i.counter]]; ok {
+		i.counter++
+		return e
+	}
+	return nil
+}
+
+type MultiStoreServer interface {
```

Comment:
I think it would be useful to comment the distinction between the different interfaces. and generally what each of them do